### ISSUE TITLE
Strong-mode analysis errors prevent DDC compilation

### ISSUE THREAD

**Issue by ochafik** [dart-lang/collection#23]
The following list of [strong-mode](https://github.com/dart-lang/dev_compiler/blob/master/STRONG_MODE.md) errors prevent code that depend on `package:collection` from being compiled with [DDC](https://github.com/dart-lang/dev_compiler).

(ran ```dartanalyzer --strong `find lib -name '*.dart'` | sort -u ``` with SDK version `1.16.0-dev.0.0`)

```
[error] Base class introduces an invalid override. The type of _DelegatingIterableBase.expand (((E) → Iterable<dynamic>) → Iterable<dynamic>) is not a subtype of Iterable<E>.expand (<T>((E) → Iterable<T>) → Iterable<T>). (./lib/src/wrappers.dart, line 194, col 24)
[error] Base class introduces an invalid override. The type of _DelegatingIterableBase.expand (((E) → Iterable<dynamic>) → Iterable<dynamic>) is not a subtype of Iterable<E>.expand (<T>((E) → Iterable<T>) → Iterable<T>). (./lib/src/wrappers.dart, line 245, col 26)
[error] Base class introduces an invalid override. The type of _DelegatingIterableBase.expand (((E) → Iterable<dynamic>) → Iterable<dynamic>) is not a subtype of Iterable<E>.expand (<T>((E) → Iterable<T>) → Iterable<T>). (./lib/src/wrappers.dart, line 97, col 25)
[error] Base class introduces an invalid override. The type of _DelegatingIterableBase.expand (((V) → Iterable<dynamic>) → Iterable<dynamic>) is not a subtype of Iterable<V>.expand (<T>((V) → Iterable<T>) → Iterable<T>). (./lib/src/wrappers.dart, line 414, col 25)
[error] Base class introduces an invalid override. The type of _DelegatingIterableBase.fold ((dynamic, (dynamic, E) → dynamic) → dynamic) is not a subtype of Iterable<E>.fold (<T>(T, (T, E) → T) → T). (./lib/src/wrappers.dart, line 194, col 24)
[error] Base class introduces an invalid override. The type of _DelegatingIterableBase.fold ((dynamic, (dynamic, E) → dynamic) → dynamic) is not a subtype of Iterable<E>.fold (<T>(T, (T, E) → T) → T). (./lib/src/wrappers.dart, line 245, col 26)
[error] Base class introduces an invalid override. The type of _DelegatingIterableBase.fold ((dynamic, (dynamic, E) → dynamic) → dynamic) is not a subtype of Iterable<E>.fold (<T>(T, (T, E) → T) → T). (./lib/src/wrappers.dart, line 97, col 25)
[error] Base class introduces an invalid override. The type of _DelegatingIterableBase.fold ((dynamic, (dynamic, V) → dynamic) → dynamic) is not a subtype of Iterable<V>.fold (<T>(T, (T, V) → T) → T). (./lib/src/wrappers.dart, line 414, col 25)
[error] Base class introduces an invalid override. The type of _DelegatingIterableBase.map (((E) → dynamic) → Iterable<dynamic>) is not a subtype of Iterable<E>.map (<T>((E) → T) → Iterable<T>). (./lib/src/wrappers.dart, line 194, col 24)
[error] Base class introduces an invalid override. The type of _DelegatingIterableBase.map (((E) → dynamic) → Iterable<dynamic>) is not a subtype of Iterable<E>.map (<T>((E) → T) → Iterable<T>). (./lib/src/wrappers.dart, line 245, col 26)
[error] Base class introduces an invalid override. The type of _DelegatingIterableBase.map (((E) → dynamic) → Iterable<dynamic>) is not a subtype of Iterable<E>.map (<T>((E) → T) → Iterable<T>). (./lib/src/wrappers.dart, line 97, col 25)
[error] Base class introduces an invalid override. The type of _DelegatingIterableBase.map (((V) → dynamic) → Iterable<dynamic>) is not a subtype of Iterable<V>.map (<T>((V) → T) → Iterable<T>). (./lib/src/wrappers.dart, line 414, col 25)
[error] Invalid override. The type of _DelegatingIterableBase.expand (((E) → Iterable<dynamic>) → Iterable<dynamic>) is not a subtype of Iterable<E>.expand (<T>((E) → Iterable<T>) → Iterable<T>). (./lib/src/wrappers.dart, line 27, col 3)
[error] Invalid override. The type of _DelegatingIterableBase.fold ((dynamic, (dynamic, E) → dynamic) → dynamic) is not a subtype of Iterable<E>.fold (<T>(T, (T, E) → T) → T). (./lib/src/wrappers.dart, line 34, col 3)
[error] Invalid override. The type of _DelegatingIterableBase.map (((E) → dynamic) → Iterable<dynamic>) is not a subtype of Iterable<E>.map (<T>((E) → T) → Iterable<T>). (./lib/src/wrappers.dart, line 54, col 3)
[error] Type check failed: const DefaultEquality() (DefaultEquality) is not of type Equality<E> (./lib/src/equality.dart, line 182, col 38)
[error] Type check failed: const DefaultEquality() (DefaultEquality) is not of type Equality<E> (./lib/src/equality.dart, line 198, col 38)
[error] Type check failed: const DefaultEquality() (DefaultEquality) is not of type Equality<E> (./lib/src/equality.dart, line 56, col 31)
[error] Type check failed: const DefaultEquality() (DefaultEquality) is not of type Equality<E> (./lib/src/equality.dart, line 99, col 53)
[error] Type check failed: const DefaultEquality() (DefaultEquality) is not of type Equality<K> (./lib/src/equality.dart, line 234, col 42)
[error] Type check failed: const DefaultEquality() (DefaultEquality) is not of type Equality<V> (./lib/src/equality.dart, line 235, col 44)
[warning] Missing concrete implementation of 'Iterable.expand', 'Iterable.fold' and 'Iterable.map' (./lib/src/unmodifiable_wrappers.dart, line 20, col 7)
[warning] Missing concrete implementation of 'Iterable.expand', 'Iterable.fold' and 'Iterable.map' (./lib/src/unmodifiable_wrappers.dart, line 93, col 7)
[warning] Missing concrete implementation of 'Iterable.expand', 'Iterable.fold' and 'Iterable.map' (./lib/src/wrappers.dart, line 194, col 7)
[warning] Missing concrete implementation of 'Iterable.expand', 'Iterable.fold' and 'Iterable.map' (./lib/src/wrappers.dart, line 245, col 7)
[warning] Missing concrete implementation of 'Iterable.expand', 'Iterable.fold' and 'Iterable.map' (./lib/src/wrappers.dart, line 339, col 7)
[warning] Missing concrete implementation of 'Iterable.expand', 'Iterable.fold' and 'Iterable.map' (./lib/src/wrappers.dart, line 414, col 7)
[warning] Missing concrete implementation of 'Iterable.expand', 'Iterable.fold' and 'Iterable.map' (./lib/src/wrappers.dart, line 97, col 7)
[warning] The redirected constructor '() → DefaultEquality' has incompatible parameters with '() → Equality<E>' (./lib/src/equality.dart, line 11, col 30)
[warning] Unsound implicit cast from (dynamic, dynamic) → int to (E, E) → int (./lib/src/priority_queue.dart, line 210, col 15)
[warning] Unsound implicit cast from (dynamic, dynamic) → int to (E, E) → int (./lib/src/priority_queue.dart, line 215, col 38)
[warning] Unsound implicit cast from Iterable<E> to List<dynamic> (./lib/src/queue_list.dart, line 38, col 25)
[warning] Unsound implicit cast from Iterable<E> to List<dynamic> (./lib/src/queue_list.dart, line 55, col 19)
[warning] Unsound implicit cast from List<dynamic> to Iterable<E> (./lib/src/queue_list.dart, line 39, col 40)
[warning] Unsound implicit cast from List<dynamic> to Iterable<E> (./lib/src/queue_list.dart, line 61, col 52)
[warning] Unsound implicit cast from List<dynamic> to Iterable<E> (./lib/src/queue_list.dart, line 67, col 52)
[warning] Unsound implicit cast from List<dynamic> to Iterable<E> (./lib/src/queue_list.dart, line 71, col 52)
[warning] Unsound implicit cast from List<dynamic> to Iterable<E> (./lib/src/queue_list.dart, line 72, col 40)
[warning] Unsound implicit cast from List<dynamic> to List<Comparable<dynamic>> (./lib/src/algorithms.dart, line 36, col 36)
[warning] Unsound implicit cast from List<dynamic> to List<Comparable<dynamic>> (./lib/src/algorithms.dart, line 84, col 34)
[warning] Unsound implicit cast from List<dynamic> to List<Iterator<dynamic>> (./lib/src/iterable_zip.dart, line 33, col 47)
[warning] Unsound implicit cast from dynamic to C (./lib/src/canonicalized_map.dart, line 61, col 11)
[warning] Unsound implicit cast from dynamic to C (./lib/src/canonicalized_map.dart, line 93, col 30)
[warning] Unsound implicit cast from dynamic to E (./lib/src/equality.dart, line 68, col 36)
[warning] Unsound implicit cast from dynamic to E (./lib/src/equality.dart, line 68, col 49)
[warning] Unsound implicit cast from dynamic to E (./lib/src/unmodifiable_wrappers.dart, line 59, col 28)
[warning] Unsound implicit cast from dynamic to E (./lib/src/unmodifiable_wrappers.dart, line 63, col 21)
[warning] Unsound implicit cast from dynamic to K (./lib/src/wrappers.dart, line 443, col 13)
[warning] Unsound implicit cast from dynamic to V (./lib/src/unmodifiable_wrappers.dart, line 151, col 41)
[warning] Unsound implicit cast from dynamic to V (./lib/src/unmodifiable_wrappers.dart, line 159, col 27)
```

**Comment by nex3** [comment]
I have some partial work locally towards making this package strong-mode clean, but last I worked on it I ran into some issues with the core libraries not being fully generic-annotated yet.

**Comment by ochafik** [comment]
Seems like strong-mode is a moving target, might be worth retrying with the latest dev sdk. And if you push your changes to a branch I'll be happy to check!

**Comment by nex3** [comment]
I've pushed my partial work to the [wip.strong-mode](https://github.com/dart-lang/collection/tree/wip.strong-mode) branch.

**Comment by ochafik** [comment]
Cool, thanks! Merged + fixed last issues in #24 :-)

**Comment by ochafik** [comment]
Fixed by https://github.com/dart-lang/collection/commit/7ac6f9b3efd8db4fccdd45e7aeec0c1df1a884a2, thanks @nex!
